test/workbox-strategies/sw/test-StrategyHandler.mjs

```diff
@@ -0,0 +1,1047 @@
+      };
+
+      const spyOne = sandbox.spy(firstPlugin, 'cacheKeyWillBeUsed');
+      const spyTwo = sandbox.spy(secondPlugin, 'cacheKeyWillBeUsed');
+
+      const initialRequest = new Request('/noPlugin');
+      const response = new Response('Test response.');
+
+      const handler = createStrategyHandler({
+        cacheName,
+        plugins: [
+          firstPlugin,
+          {}, // Intentionally empty to ensure it's filtered out.
+          secondPlugin,
+        ],
+      });
+
+      await handler.cachePut(initialRequest, response);
+
+      expect(spyOne.calledOnceWith(sinon.match({
+        mode: 'write',
+        request: initialRequest,
+      }))).to.be.true;
+      expect(spyOne.thisValues[0]).to.eql(firstPlugin);
+
+      expect(spyTwo.calledOnceWith(sinon.match({
+        mode: 'write',
+        request: firstPluginReturnValue,
+      }))).to.be.true;
+      expect(spyTwo.thisValues[0]).to.eql(secondPlugin);
+
+      expect(cachePutStub.calledOnce).to.be.true;
+      // Check the url of the Request passed to cache.put().
+      expect(cachePutStub.args[0][0].url).to.eql(`${self.location.origin}/secondPlugin`);
+      expect(cachePutStub.args[0][1]).to.eql(response);
+    });
+
+    it(`should call the quota exceeded callbacks when there's a QuotaExceeded error`, async function() {
+      const callback1 = sandbox.stub();
+      registerQuotaErrorCallback(callback1);
+      const callback2 = sandbox.stub();
+      registerQuotaErrorCallback(callback2);
+
+      const cacheName = 'test-cache';
+      const testCache = await caches.open(cacheName);
+      sandbox.stub(self.caches, 'open').returns(Promise.resolve(testCache));
+      sandbox.stub(testCache, 'put').throws('QuotaExceededError');
+
+      const handler = createStrategyHandler({cacheName});
+
+      try {
+        await handler.cachePut('ignored', new Response());
+        throw new Error('Unexpected success.');
+      } catch (error) {
+        expect(error.name).to.eql('QuotaExceededError');
+      }
+      expect(callback1.calledOnce).to.be.true;
+      expect(callback2.calledOnce).to.be.true;
+    });
+
+    it(`should not call the quota exceeded callbacks when there's a non-QuotaExceeded error`, async function() {
+      const callback = sandbox.stub();
+      registerQuotaErrorCallback(callback);
+
+      const cacheName = 'test-cache';
+      const testCache = await caches.open(cacheName);
+      sandbox.stub(self.caches, 'open').returns(Promise.resolve(testCache));
+      sandbox.stub(testCache, 'put').throws('NetworkError');
+
+      const handler = createStrategyHandler({cacheName});
+
+      try {
+        await handler.cachePut('ignored', new Response());
+        throw new Error('Unexpected success.');
+      } catch (error) {
+        expect(error.name).to.eql('NetworkError');
+      }
+      expect(callback.called).to.be.false;
+    });
+  });
+
+  // These tests were copied from `test-cachWerapper` to ensure we don't
```

Comment:
"cachWerapper" is a classic Macbook Pro keyboard typo.

Reply:
Yup, fixed!